File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> need;
        unordered_map<char, int> wind;
        for(char c : t){
            need[c]++;
        }
        int req = need.size();
        int form = 0;
        int minlen = INT_MAX;
        int left = 0;
        int start = 0;
        for(int i=0; i<s.size(); i++){
            char c = s[i];
            wind[c]++;
            if(need.count(c) && wind[c] == need[c]) form++;
            while(left <= i && form == req){
                if(i-left+1 < minlen){
                    minlen = i - left + 1;
                    start = left;
                }
                char ch = s[left];
                wind[ch]--;
                if(need.count(ch) && wind[ch] < need[ch]) form--;
                left++;
            }
        }
        if(minlen == INT_MAX) return "";
        return s.substr(start, minlen);
    }
};

File: 76-minimum-window-substring/minimum-window-substring.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t or len(s) < len(t):
            return ""

        ans = ""
        need = Counter(t)
        wind = {}

        req = len(need)
        form = 0

        left = 0
        min_len = float('inf')

        for i in range(len(s)):
            ch = s[i]
            wind[ch] = wind.get(ch, 0) + 1

            if ch in need and wind[ch] == need[ch]:
                form += 1

            while form == req:

                if i - left + 1 < min_len:
                    ans = s[left:i + 1]
                    min_len = i - left + 1

                left_ch = s[left]

                wind[left_ch] -= 1
                if left_ch in need and wind[left_ch] < need[left_ch]:
                    form -= 1

                left += 1

        return ans